saleor/payment/utils.py

```diff
@@ -125,9 +125,78 @@ def clean_authorize(payment: Payment):
         raise PaymentError('Charged transactions cannot be authorized again.')
 
 
+def call_gateway(
+        func_name, transaction_kind, payment, payment_token, **extra_params):
+    """Helper that calls the passed gateway function and handles exceptions.
+
+    Additionally does validation of the returned gateway response."""
+    gateway, gateway_params = get_payment_gateway(payment.gateway)
+    gateway_response = None
+
+    #TODO: improve so only one argument is sent (dict?) instead of passing
+    # multiple kwargs, eventually keep connection params separate
+    #TODO: add additional information to payment data (e.g. shipping/billing)
+    try:
+        gateway_response = getattr(gateway, func_name)(
+            payment=model_to_dict(payment), payment_token=payment_token,
+            **extra_params, **gateway_params)
+        validate_gateway_response(gateway_response)
+    except AttributeError:
+        logger.exception('Gateway doesn\'t implement %s', func_name)
+    except GatewayError:
+        logger.exception('Gateway response validation failed')
+        gateway_response = None  # set response none as the validation failed
+    except Exception:
+        logger.exception('Gateway encountered an error')
+    finally:
+        transaction = create_transaction(
+            payment=payment,
+            kind=transaction_kind,
+            payment_token=payment_token,
+            gateway_response=gateway_response,
+            **extra_params)
+
+    if not transaction.is_success:
+        # attempt to get errors from response, if none raise a generic one
+        raise PaymentError(gateway_response.get(
+            'errors', GENERIC_TRANSACTION_ERROR))
+
+    return transaction
+
+
+def validate_gateway_response(response):
+    """Validates response to be a correct format for Saleor to process."""
+    if not isinstance(response, dict):
+        raise GatewayError('Gateway needs to return a dictionary')
+    required_fields = {'transaction_id', 'is_success', 'errors'}
```

Comment:
It's not required for Saleor to work, you can send it, yes, but I wouldn't go as far as to make it a required field. Also you can send back whatever you want, the whole response returned from gateway is saved, not a specific key from the dictionary, this makes it easier as you can also return additional data and not worry about proper key naming. Notice I'm not raising a validation error if there's more returned than those 3 keys. Just as long as it's encodable by DjangoJsonEncoder, which I should probably check here just in case.

Reply:
If required means required for Saleor to work, I can accept it. :D

Reply:
I might've used wrong wording, but the `required_keys` is something that, in my mind, is required in order to process payments, `raw_response` is not needed in order to do.

Reply:
Just to elaborate it, raw response is not crucial to saleor but it is essential to a payment system. If someone write a provider without saving the raw response, I probably won't use it. The payment gateway is related with real money, and we should take it seriously. Without raw response, we can barely debug the transaction issues if any issues happens and this is a big problem.